usecases/internal/validation.go

```diff
@@ -0,0 +1,350 @@
+// from a slice, or validate all items with detailed error reporting.
+type Validator[T any] interface {
+	Validate(item T) error
+	ValidateFirst(items []T) (T, error)
+	ValidateAll(items []T) ([]T, []error)
+	WithName(name string) Validator[T]
+}
+
+// ValidationRule is a function that validates a specific type.
+//
+// Rules are composable and can be combined using CombineRules or
+// applied conditionally using ConditionalRule.
+type ValidationRule[T any] func(T) error
+
+// BaseValidator provides common validation functionality for any type.
+//
+// It implements the Validator interface and supports method chaining
+// for building complex validators with multiple rules.
+type BaseValidator[T any] struct {
+	rules []ValidationRule[T]
+	name  string
+}
+
+// NewValidator creates a new validator for type T.
+//
+// Example:
+//
+//	validator := NewValidator[*model.MeasurementDataType]().
+//		WithName("Power Validator").
+//		WithRule(RequireMeasurementId()).
+//		WithRule(RequireMeasurementValue())
+//
+//	err := validator.Validate(measurement)
+func NewValidator[T any]() *BaseValidator[T] {
+	return &BaseValidator[T]{}
+}
+
+// WithName sets the validator name for better error messages.
+//
+// The name appears in error messages to help identify which validator failed.
+//
+// Example:
+//
+//	validator.WithName("Power Measurement")
+//	// Error: "Power Measurement validation failed at rule 2: Value is required"
+func (v *BaseValidator[T]) WithName(name string) *BaseValidator[T] {
+	v.name = name
+	return v
+}
+
+// WithRule adds a validation rule to the validator.
+//
+// Rules are executed in the order they are added. The first rule that fails
+// stops validation and returns an error.
+//
+// Example:
+//
+//	validator.WithRule(RequireField(getter, "FieldName")).
+//		WithRule(ValidateRange(getter, 0, 100, "Value"))
+func (v *BaseValidator[T]) WithRule(rule ValidationRule[T]) *BaseValidator[T] {
+	v.rules = append(v.rules, rule)
+	return v
+}
+
+// Validate applies all rules to a single item.
+//
+// Returns nil if all rules pass, or the first error encountered.
+// Error messages include the validator name (if set) and rule position.
+func (v *BaseValidator[T]) Validate(item T) error {
+	for i, rule := range v.rules {
+		if err := rule(item); err != nil {
+			if v.name != "" {
+				return fmt.Errorf("%s validation failed at rule %d: %w", v.name, i+1, err)
+			}
+			return fmt.Errorf("validation failed at rule %d: %w", i+1, err)
+		}
+	}
+	return nil
+}
+
+// ValidateFirst returns the first valid item from a slice.
+//
+// This is useful when you have multiple items but only need one valid one.
+// Returns the first item that passes all validation rules.
+//
+// Example:
+//
+//	measurements := []model.MeasurementDataType{...}
+//	valid, err := validator.ValidateFirst(measurements)
+//	if err != nil {
+//		return api.ErrDataNotAvailable
+//	}
+func (v *BaseValidator[T]) ValidateFirst(items []T) (T, error) {
```

Comment:
I dislike the name of this function. The function is not validating the first element of items[], it's finding the first valid element.
```suggestion
func (v *BaseValidator[T]) FindFirstValidItem(items []T) (T, error) {
```